File: PHP/src/OneLang/Test/TestCases/BasicTests.php
<?php

namespace OneLang\Test\TestCases\BasicTests;

use OneLang\Test\TestCase\ITestCollection;
use OneLang\Test\TestCase\SyncTestCase;
use OneLang\Test\TestCase\TestCase;

class BasicTests implements ITestCollection {
    public $name = "BasicTests";
    
    function printsOneLineString() {
        \OneLang\Core\console::log("Hello World!");
    }
    
    function printsMultiLineString() {
        \OneLang\Core\console::log("Hello\nWorld!");
    }
    
    function printsTwoStrings() {
        \OneLang\Core\console::log("Hello");
        \OneLang\Core\console::log("HelloWorld!");
    }
    
    function printsEscapedString() {
        \OneLang\Core\console::log("dollar: \$");
        \OneLang\Core\console::log("backslash: \\");
        \OneLang\Core\console::log("newline: \n");
        \OneLang\Core\console::log("escaped newline: \\n");
        \OneLang\Core\console::log("dollar after escape: \\\$");
    }
    
    function printsEscapedTemplateString() {
        \OneLang\Core\console::log("dollar: \$");
        \OneLang\Core\console::log("backslash: \\");
        \OneLang\Core\console::log("newline: \n");
        \OneLang\Core\console::log("escaped newline: \\n");
        \OneLang\Core\console::log("dollar after escape: \\\$");
    }
    
    function regexReplace() {
        \OneLang\Core\console::log(preg_replace("/\\$/", "x", "a\$b\$c"));
        \OneLang\Core\console::log(preg_replace("/x/", "$", "Test1: xx"));
        \OneLang\Core\console::log(preg_replace("/y/", "\\\\", "Test2: yy"));
        \OneLang\Core\console::log(preg_replace("/z/", "\\\\$", "Test3: zz"));
    }
    
    function json() {
        \OneLang\Core\console::log(json_encode(null, JSON_UNESCAPED_SLASHES));
        \OneLang\Core\console::log(json_encode(true, JSON_UNESCAPED_SLASHES));
        \OneLang\Core\console::log(json_encode(false, JSON_UNESCAPED_SLASHES));
        \OneLang\Core\console::log(json_encode("string", JSON_UNESCAPED_SLASHES));
        \OneLang\Core\console::log(json_encode(0.123, JSON_UNESCAPED_SLASHES));
        \OneLang\Core\console::log(json_encode(123, JSON_UNESCAPED_SLASHES));
        \OneLang\Core\console::log(json_encode(123.456, JSON_UNESCAPED_SLASHES));
        \OneLang\Core\console::log(json_encode(Array("a" => "b"), JSON_UNESCAPED_SLASHES));
        \OneLang\Core\console::log(json_encode("\$", JSON_UNESCAPED_SLASHES));
        \OneLang\Core\console::log(json_encode("A \\ B", JSON_UNESCAPED_SLASHES));
        \OneLang\Core\console::log(json_encode("A \\\\ B", JSON_UNESCAPED_SLASHES));
    }
    
    function phpGeneratorBugs() {
        \OneLang\Core\console::log("Step1: " . "A \$ B");
        \OneLang\Core\console::log("Step2: " . json_encode("A \$ B", JSON_UNESCAPED_SLASHES));
        \OneLang\Core\console::log("Step3: " . preg_replace("/\\$/", "\\\\$", json_encode("A \$ B", JSON_UNESCAPED_SLASHES)));
        \OneLang\Core\console::log("Step3 w/o JSON: " . preg_replace("/\\$/", "\\\\$", "A \$ B"));
    }
    
    function getTestCases() {
        return array(new SyncTestCase("PrintsOneLineString", function ($_) { return $this->printsOneLineString(); }), new SyncTestCase("PrintsMultiLineString", function ($_) { return $this->printsMultiLineString(); }), new SyncTestCase("PrintsTwoStrings", function ($_) { return $this->printsTwoStrings(); }), new SyncTestCase("PrintsEscapedString", function ($_) { return $this->printsEscapedString(); }), new SyncTestCase("PrintsEscapedTemplateString", function ($_) { return $this->printsEscapedTemplateString(); }), new SyncTestCase("RegexReplace", function ($_) { return $this->regexReplace(); }), new SyncTestCase("Json", function ($_) { return $this->json(); }), new SyncTestCase("PhpGeneratorBugs", function ($_) { return $this->phpGeneratorBugs(); }));
    }
}

File: Python/OneLang/Test/TestCases/BasicTests.py
from onelang_core import *
import OneLang.Test.TestCase as testCase
import re
import json

class BasicTests:
    def __init__(self):
        self.name = "BasicTests"
    
    def prints_one_line_string(self):
        console.log("Hello World!")
    
    def prints_multi_line_string(self):
        console.log("Hello\nWorld!")
    
    def prints_two_strings(self):
        console.log("Hello")
        console.log("HelloWorld!")
    
    def prints_escaped_string(self):
        console.log("dollar: $")
        console.log("backslash: \\")
        console.log("newline: \n")
        console.log("escaped newline: \\n")
        console.log("dollar after escape: \\$")
    
    def prints_escaped_template_string(self):
        console.log(f'''dollar: $''')
        console.log(f'''backslash: \\''')
        console.log(f'''newline: \n''')
        console.log(f'''escaped newline: \\n''')
        console.log(f'''dollar after escape: \\$''')
    
    def regex_replace(self):
        console.log(re.sub("\\$", "x", "a$b$c"))
        console.log(re.sub("x", "$", "Test1: xx"))
        console.log(re.sub("y", "\\\\", "Test2: yy"))
        console.log(re.sub("z", "\\\\$", "Test3: zz"))
    
    def json(self):
        console.log(json.dumps(None, separators=(',', ':')))
        console.log(json.dumps(True, separators=(',', ':')))
        console.log(json.dumps(False, separators=(',', ':')))
        console.log(json.dumps("string", separators=(',', ':')))
        console.log(json.dumps(0.123, separators=(',', ':')))
        console.log(json.dumps(123, separators=(',', ':')))
        console.log(json.dumps(123.456, separators=(',', ':')))
        console.log(json.dumps({
            "a": "b"
        }, separators=(',', ':')))
        console.log(json.dumps("$", separators=(',', ':')))
        console.log(json.dumps("A \\ B", separators=(',', ':')))
        console.log(json.dumps("A \\\\ B", separators=(',', ':')))
    
    def php_generator_bugs(self):
        console.log("Step1: " + "A $ B")
        console.log("Step2: " + json.dumps("A $ B", separators=(',', ':')))
        console.log("Step3: " + re.sub("\\$", "\\\\$", json.dumps("A $ B", separators=(',', ':'))))
        console.log("Step3 w/o JSON: " + re.sub("\\$", "\\\\$", "A $ B"))
    
    def get_test_cases(self):
        return [testCase.SyncTestCase("PrintsOneLineString", lambda _: self.prints_one_line_string()), testCase.SyncTestCase("PrintsMultiLineString", lambda _: self.prints_multi_line_string()), testCase.SyncTestCase("PrintsTwoStrings", lambda _: self.prints_two_strings()), testCase.SyncTestCase("PrintsEscapedString", lambda _: self.prints_escaped_string()), testCase.SyncTestCase("PrintsEscapedTemplateString", lambda _: self.prints_escaped_template_string()), testCase.SyncTestCase("RegexReplace", lambda _: self.regex_replace()), testCase.SyncTestCase("Json", lambda _: self.json()), testCase.SyncTestCase("PhpGeneratorBugs", lambda _: self.php_generator_bugs())]